### PR TITLE
Vault :allow ratelimit and client timeout to be configurable

### PR DESCRIPTION
**Pull request by poornas**
## Motivation and Context

Allow vault client timeout and rate limit to be configurable as an advanced option.
## Regression
 No

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/environment.go [hunk 1]
```diff
@@ -160,6 +186,19 @@ func (env environment) LookupKMSConfig(config crypto.KMSConfig) (err error) {
 	return nil
 }
 
+// parses rate limit value in the format rate[:burst] where :burst is optional
+func parseRateLimit(val string) (rate float64, burst int, err error) {
+	_, err = fmt.Sscanf(val, "%f:%d", &rate, &burst)
+	if err != nil {
+		rate, err = strconv.ParseFloat(val, 64)
+		if err != nil {
+			err = fmt.Errorf("%v was provided but incorrectly formatted")
```
Comment (harshavardhana) [inline cmd/environment.go]: %v should be used with an arg

```
Errorf format %v reads arg #1, but call has 0 args
1639
[0mMakefile:27: recipe for target 'vet' failed
1640
[91mmake: *** [vet] Error 2
```

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-worm.sh | :x: [more...](#5a439e4e-6c2f-47d9-a339-62c4e188a0a4) |
<a name="5a439e4e-6c2f-47d9-a339-62c4e188a0a4"></a>
### 7470-cf734bf/mint-worm.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:32317
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            worm
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7b45fd7a0ca5:/mint/log /tmp/mint-logs'

(1/1) Running worm tests ... FAILED in 3 seconds
{
  "alert": "",
  "args": {
    "bucketName": "bucket-1ttqi0j8ak22gvigiw43z42l02fggp4xnzhy529glsobgyld9k5zvog",
    "expiry": 60000000000,
    "objectName": "testObject"
  },
  "duration": 3055,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:32317/bucket-1ttqi0j8ak22gvigiw43z42l02fggp4xnzhy529glsobgyld9k5zvog: dial tcp 72.28.97.57:32317: connect: no route to host",
  "function": "PutAndDelete",
  "message": "WORM_MODE ON  - CreateBucket Failed",
  "name": "test worm mode",
  "status": "FAIL"
}

Executed 0 out of 1 tests successfully.
```

File: docs/kms/README.md [hunk 2]
```diff
@@ -108,6 +108,15 @@ export MINIO_SSE_VAULT_NAMESPACE=ns1
 
 Note: If [Vault Namespaces](https://learn.hashicorp.com/vault/operations/namespaces) are in use, MINIO_SSE_VAULT_NAMESPACE variable needs to be set before setting approle and transit secrets engine.
 
+Advanced options:
+MINIO_SSE_VAULT_CLIENT_TIMEOUT can be set to a custom timeout, defaults to 60 seconds.
+MINIO_SSE_VAULT_RATE_LIMIT limits the rate at which requests are sent to Vault,and can be specified in the format rate:[burst] where rate and burst are "operations per second". If burst is unspecified, it defaults to rate. 
+```
+export MINIO_SSE_VAULT_CLIENT_TIMEOUT=50  # timeout duration in seconds
+export MINIO_SSE_VAULT_RATE_LIMIT=1000:10 # rate limit to 50 requests, with burst limit of 10
```
Comment (kannappanr) [inline docs/kms/README.md]: Is this a typo? Should `MINIO_SSE_VAULT_RATE_LIMIT=1000:10 # rate limit to 50 requests` be `MINIO_SSE_VAULT_RATE_LIMIT=1000:10 # rate limit to 1000 requests`

File: cmd/environment.go (hunk 1 above)
Comment (poornas) [reply cmd/environment.go]: fixed

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#0bdb2224-0a27-4041-9bac-d1b3693ea39a) |
| mint-compression-dist-xl.sh | :x: [more...](#1f4660ba-88cc-4ea7-b46e-37bcd928c745) |
| mint-dist-xl.sh | :x: [more...](#9bd80fa4-b135-4279-b547-13d5fc15d78a) |
<a name="1f4660ba-88cc-4ea7-b46e-37bcd928c745"></a>
### 7470-4fed006/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:30349
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 628aeb2a33f3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 3 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-rvvn33vy69ze91",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 3055,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.57:30349/aws-sdk-go-test-rvvn33vy69ze91: dial tcp 72.28.97.57:30349: connect: no route to host",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="0bdb2224-0a27-4041-9bac-d1b3693ea39a"></a>
### 7470-4fed006/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31277
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7a04f863ec03:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 4 minutes and 5 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 1 minutes and 13 seconds
(4/14) Running aws-sdk-ruby tests ... done in 14 seconds
(5/14) Running awscli tests ... done in 1 minutes and 26 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 1 minutes and 8 seconds
{
  "name": "mc",
  "duration": "65494",
  "function": "test_put_object_with_storage_class",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-3731 --quiet --no-color cp --storage-class REDUCED_REDUNDANCY /mint/data/datafile-1-MB myminio/mc-test-bucket-18127/mc-test-object-13596  >>> \n`/mint/data/datafile-1-MB` -> `myminio/mc-test-bucket-18127/mc-test-object-13596`\nmc: <ERROR> Failed to copy `/mint/data/datafile-1-MB`. Multiple disks failures, unable to write data.\nmc: <ERROR> Session safely terminated. To resume session `mc session resume pHNRNbQY`"
}

Executed 6 out of 14 tests successfully.
```
<a name="9bd80fa4-b135-4279-b547-13d5fc15d78a"></a>
### 7470-4fed006/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:30250
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4b89e844ebb8:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 6 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 4 seconds
(5/14) Running awscli tests ... done in 2 minutes and 56 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 18 seconds
{
  "name": "mc",
  "duration": "7227",
  "function": "test_put_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-7588 --quiet --no-color rm myminio/mc-test-bucket-4001/mc-test-object-11634  >>> \nmc: <ERROR> Unable to list folder. Get http://72.28.97.53:30250/mc-test-bucket-4001/?location=: dial tcp 72.28.97.53:30250: connect: no route to host\nmc: <ERROR> Failed to remove `myminio/mc-test-bucket-4001/mc-test-object-11634`. "
}

Executed 6 out of 14 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#8d7a68d3-aff2-47b8-aa2a-bdc9a9919728) |
| mint-compression-dist-xl.sh | :x: [more...](#e8a71b95-a628-4371-8e93-8f3bf7ce1bc2) |
| mint-compression-fs.sh | :x: [more...](#7d017409-3779-40f8-9a58-4d9e3d031722) |
| mint-dist-xl.sh | :x: [more...](#7f460aca-d00e-43ca-a549-961f4bf56c89) |
<a name="7f460aca-d00e-43ca-a549-961f4bf56c89"></a>
### 7470-e591988/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:30759
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 76df5fd14042:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... FAILED in 8 minutes and 41 seconds
{
  "name": "awscli",
  "duration": 8302,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioWriteQuorum) when calling the CreateBucket operation (reached max retries: 4): Multiple disks failures, unable to write data."
}

Executed 4 out of 14 tests successfully.
```
<a name="e8a71b95-a628-4371-8e93-8f3bf7ce1bc2"></a>
### 7470-e591988/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.55:32356
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b74c260d8d87:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 1 minutes and 50 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 27 seconds
(8/14) Running minio-dotnet tests ... done in 32 seconds
(9/14) Running minio-go tests ... done in 34 seconds
(10/14) Running minio-java tests ... FAILED in 58 seconds
{
  "name": "minio-java",
  "function": "listIncompleteUploads(final String bucketName, final String prefix, final boolean recursive)",
  "args": "prefix: minio, recursive: true",
  "duration": 324,
  "status": "FAIL",
  "error": "java.net.NoRouteToHostException: No route to host (Host unreachable) >>> [java.net.PlainSocketImpl.socketConnect(Native Method), java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350), java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206), java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188), java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392), java.net.Socket.connect(Socket.java:589), okhttp3.internal.platform.Platform.connectSocket(Platform.java:130), okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:247), okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167), okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:257), okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135), okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:264), okhttp3.RealCall.execute(RealCall.java:93), io.minio.MinioClient.executeReq(MinioClient.java:1097), io.minio.MinioClient.execute(MinioClient.java:1056), io.minio.MinioClient.executeDelete(MinioClient.java:1333), io.minio.MinioClient.abortMultipartUpload(MinioClient.java:4899), io.minio.MinioClient.putObject(MinioClient.java:4087), io.minio.MinioClient.putObject(MinioClient.java:3570), FunctionalTest.listIncompleteUploads_test3(FunctionalTest.java:1682), FunctionalTest.runTests(FunctionalTest.java:2993), FunctionalTest.main(FunctionalTest.java:3121)]"
}

Executed 9 out of 14 tests successfully.
```
<a name="8d7a68d3-aff2-47b8-aa2a-bdc9a9919728"></a>
### 7470-e591988/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:30558
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 1b91d25f372f:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 30 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-k96lmtwv2rdcrz",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 390010,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.60:30558/aws-sdk-go-test-k96lmtwv2rdcrz: dial tcp 72.28.97.60:30558: i/o timeout",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="7d017409-3779-40f8-9a58-4d9e3d031722"></a>
### 7470-e591988/mint-compression-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.53:32680
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3083a91f778b:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 45 seconds
(4/14) Running aws-sdk-ruby tests ... done in 3 seconds
(5/14) Running awscli tests ... done in 47 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 17 seconds
(8/14) Running minio-dotnet tests ... done in 27 seconds
(9/14) Running minio-go tests ... done in 43 seconds
(10/14) Running minio-java tests ... done in 1 minutes and 52 seconds
(11/14) Running minio-js tests ... done in 38 seconds
(12/14) Running minio-py tests ... FAILED in 4 minutes and 17 seconds
{
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1220, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1885, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1228, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "duration": 1323,
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "status": "FAIL",
  "message": "AccessDenied: message: Access Denied",
  "args": {
    "object_name": "682a6b5f-5d36-446e-bcec-cb9a0d3ad2ee",
    "bucket_name": "minio-py-test-347e0922-7bd6-4c71-81f4-a69adb0ad523"
  }
}

Executed 11 out of 14 tests successfully.
```

**Comment by poornas** [comment]
Closing this as underlying vault client env variables can be used instead

**Pull request closed without merging**